src/ring_theory/dedekind_domain/factorization.lean

```diff
@@ -0,0 +1,196 @@
+/-
+Copyright (c) 2022 María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: María Inés de Frutos-Fernández
+-/
+import ring_theory.dedekind_domain.ideal
+/-!
+# Factorization of ideals of Dedekind domains
+Every nonzero ideal `I` of a Dedekind domain `R` can be factored as a product `∏_v v^{n_v}` over the
+maximal ideals of `R`, where the exponents `n_v` are natural numbers.
+
+TODO: Extend the results in this file to fractional ideals of `R`.
+
+## Main results
+- `ideal.finite_factors` : Only finitely many maximal ideals of `R` divide a given nonzero ideal.
+- `ideal.factorization` : The ideal `I` equals the finprod `∏_v v^(val_v(I))`, where `val_v(I)`
+  denotes the multiplicity of `v` in the factorization of `I` and `v` runs over the maximal
+  ideals of `R`.
+
+## Tags
+dedekind domain, ideal, factorization
+-/
+
+noncomputable theory
+open_locale big_operators classical
+
+open set function unique_factorization_monoid
+
+/-! ### Factorization of ideals of Dedekind domains -/
+
+open is_dedekind_domain
+
+variables {R : Type*} [comm_ring R] [is_domain R] [is_dedekind_domain R] {K : Type*} [field K]
+  [algebra R K] [is_fraction_ring R K] (v : height_one_spectrum R)
+
+/-- Only finitely many maximal ideals of `R` divide a given nonzero ideal. -/
+lemma ideal.finite_factors {I : ideal R} (hI : I ≠ 0) :
+  { v : height_one_spectrum R | v.as_ideal ∣ I }.finite :=
+begin
+  haveI h_fin := fintype_subtype_dvd I hI,
+  let f' : finset (ideal R) := finset.map
+    ⟨(λ J : {x // x ∣ I}, J.val), subtype.coe_injective⟩ h_fin.elems,
+  have h_eq : { v : height_one_spectrum R | v.as_ideal ∣ I } =
+    { v : height_one_spectrum R | v.as_ideal ∈ f' },
+  { ext v,
+    simp_rw [finset.mem_map, exists_prop, subtype.exists, embedding.coe_fn_mk,
+      exists_and_distrib_right, exists_eq_right],
+    exact ⟨λ h, ⟨h, fintype.complete ⟨v.as_ideal, h⟩⟩, λ ⟨hv, h⟩, hv⟩, },
+  rw h_eq,
+  have hv_inj : injective (λ (v : height_one_spectrum R), v.as_ideal),
+  { intros v w hvw,
+    exact (height_one_spectrum.ext_iff v w).mpr hvw, },
+  exact finite.preimage_embedding ⟨(λ v : height_one_spectrum R, v.as_ideal), hv_inj⟩
+    (finite_mem_finset (f')),
```

Comment:
Here's a slightly more direct way to do this proof, avoiding going through multiple finsets.
```suggestion
  rw [← set.finite_coe_iff, set.coe_set_of],
  haveI h_fin := fintype_subtype_dvd I hI,
  refine finite.of_injective (λ v, (⟨(v : height_one_spectrum R).as_ideal, v.2⟩ : {x // x ∣ I})) _,
  intros v w hvw,
  simp only at hvw,
  exact subtype.coe_injective ((height_one_spectrum.ext_iff ↑v ↑w).mpr hvw)
```

Reply:
We might even want a lemma saying something like `injective f → finite {y // p y} → finite {x // p (f x)}`.